x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java

```diff
@@ -0,0 +1,131 @@
+ *
+ * It holds other services used for this task to get relevance and curations settings
+ */
+public class RelevanceMatchQueryRewriter {
+
+    private final RelevanceSettingsService relevanceSettingsService;
+    private final CurationsService curationsService;
+
+    private final QueryFieldsResolver queryFieldsResolver;
+
+    public RelevanceMatchQueryRewriter(
+        RelevanceSettingsService relevanceSettingsService,
+        CurationsService curationsService,
+        QueryFieldsResolver queryFieldsResolver
+    ) {
+        this.relevanceSettingsService = relevanceSettingsService;
+        this.curationsService = curationsService;
+        this.queryFieldsResolver = queryFieldsResolver;
+    }
+
+    private static QueryBuilder applyHiddenDocs(QueryBuilder queryBuilder, CurationSettings curationSettings) {
+        if (curationSettings.hiddenDocs().isEmpty() == false) {
+            BoolQueryBuilder booleanQueryBuilder = new BoolQueryBuilder();
+            booleanQueryBuilder.should(queryBuilder);
+
+            BoolQueryBuilder hiddenDocsBuilder = new BoolQueryBuilder();
+            for (CurationSettings.DocumentReference hiddenDoc : curationSettings.hiddenDocs()) {
+                BoolQueryBuilder mustQueryBuilder = new BoolQueryBuilder();
+                mustQueryBuilder.must(new TermsQueryBuilder("_index", hiddenDoc.index()));
+                mustQueryBuilder.must(new TermsQueryBuilder("_id", hiddenDoc.id()));
+
+                hiddenDocsBuilder.mustNot(mustQueryBuilder);
+            }
+            booleanQueryBuilder.filter(hiddenDocsBuilder);
+
+            queryBuilder = booleanQueryBuilder;
+        }
+        return queryBuilder;
+    }
+
+    private static QueryBuilder applyPinnedDocs(QueryBuilder queryBuilder, CurationSettings curationSettings) {
+        if (curationSettings.pinnedDocs().isEmpty() == false) {
+            final PinnedQueryBuilder.Item[] items = curationSettings.pinnedDocs()
+                .stream()
+                .map(docRef -> new PinnedQueryBuilder.Item(docRef.index(), docRef.id()))
+                .toArray(PinnedQueryBuilder.Item[]::new);
+            queryBuilder = new PinnedQueryBuilder(queryBuilder, items);
+        }
+        return queryBuilder;
+    }
+
+    public Query rewriteQuery(RelevanceMatchQueryBuilder relevanceMatchQueryBuilder, SearchExecutionContext context) throws IOException {
+        Map<String, Float> fieldsAndBoosts = retrieveFieldsAndBoosts(relevanceMatchQueryBuilder.getRelevanceSettingsId(), context);
+        final QueryBuilder combinedFieldsBuilder = new CombinedFieldsQueryBuilder(relevanceMatchQueryBuilder.getQuery(), fieldsAndBoosts);
+
+        return applyCurations(combinedFieldsBuilder, relevanceMatchQueryBuilder).toQuery(context);
+    }
+
+    private Map<String, Float> retrieveFieldsAndBoosts(String relevanceSettingsId, SearchExecutionContext context) {
+        Map<String, Float> fieldsAndBoosts;
+        if (relevanceSettingsId != null) {
```

Comment:
Consider checking for null/empty here instead of just null?

Reply:
I generally like to fail as quickly as possible, which would mean keeping the checks in the constructor, but in this case that's a strong opinion very weakly held. 🙂